common/src/ray_tracing.rs

```diff
@@ -0,0 +1,105 @@
+/// normals are given in. Specifically, the `start_node_transform` matrix converts this coordinate system to node coordinates.
+///
+/// The `tanh_distance` is the hyperbolic tangent of the distance along the ray to check for intersections.
+pub fn trace_ray(
+    start_node_transform: na::Matrix4<f32>,
+    ray: &Ray,
+    tanh_distance: f32,
+) -> RayTracingResult {
+    // A collision check is assumed to be a miss until a collision is found.
+    // This `result` variable gets updated over time before being returned.
+    let mut result = RayTracingResult {
+        tanh_distance,
+        intersection: None,
+    };
+
+    let node_ray = start_node_transform * ray;
+
+    // As a temporary placeholder implementation, check for collision against the z=0 plane going in the -z direction.
+
+    const EPSILON: f32 = 1e-4;
+    let found_collision = if node_ray.direction.z >= 0.0 {
+        // Going the wrong way. Don't report a collision.
+        false
+    } else if (-EPSILON..=0.0).contains(&node_ray.position.z) {
+        // Extra logic to ensure precision issues don't allow a collider to clip through a surface
+        result.tanh_distance = 0.0;
+        true
+    } else {
+        // Compute needed tanh_distance to reach z=0 plane.
+        let tanh_distance = -node_ray.position.z / node_ray.direction.z;
+        if tanh_distance >= 0.0 && tanh_distance < result.tanh_distance {
+            // Report a collision if the needed tanh_distance is in the right range
+            result.tanh_distance = tanh_distance;
+            true
+        } else {
+            false
+        }
+    };
+
+    if found_collision {
+        result.intersection = Some(RayTracingIntersection {
+            normal: math::mtranspose(&start_node_transform) * na::Vector4::z(),
+        });
+    }
+
+    result
+}
+
+/// Contains all information that was discovered as a result of ray tracing.
+pub struct RayTracingResult {
+    /// The tanh of the length of the resulting ray segment so far. As new intersections are found, the
+    /// ray segment gets shorter each time.
+    pub tanh_distance: f32,
+
+    /// Information about the intersection at the end of the ray segment. If this is `None`, there
+    /// are no intersections.
+    pub intersection: Option<RayTracingIntersection>,
+}
+
+/// Information about the intersection at the end of a ray segment.
+pub struct RayTracingIntersection {
+    /// The normal vector of the ray tracing intersection surface in the original coordinate system
+    /// of the ray tracing
+    pub normal: na::Vector4<f32>,
+}
+
+/// A ray in hyperbolic space. The fields must be lorentz normalized, with `mip(position, position) == -1`,
+/// `mip(direction, direction) == 1`, and `mip(position, direction) == 0`.
+pub struct Ray {
+    pub position: na::Vector4<f32>,
+    pub direction: na::Vector4<f32>,
+}
+
+impl Ray {
+    pub fn new(position: na::Vector4<f32>, direction: na::Vector4<f32>) -> Ray {
```

Comment:
Possibly, although I don't think we `debug_assert` most normalization invariants in the rest of the codebase.

Reply:
Might not be a bad time to start.

Reply:
I'm having second thoughts about implementing this.
* Approximate equality is needed for this debug assertion.
* `approx` is currently a dev-dependency, so it only works in tests.
* `approx` has no `debug_assert_abs_diff_eq!` method, so I would need to use `debug_assert!(abs_diff_eq!(...))` instead, making error messages less useful unless additional boilerplate code is added.
* I currently highly doubt this will be a common source of bugs. Trying to make this more airtight might be better done when we start having separate types for various normalization states as you commented below.

If you would still prefer me to add this, which approach would you recommend. Should I add the dependency, add new boilerplate code somewhere (where would you recommend?), or just do something like the following?
```rust
debug_assert!(
    (math::mip(&position, &position) + 1.0).abs() < 1e-5,
    "mip(position, position) == {}",
    math::mip(&position, &position)
);
```

Reply:
I had imagined we'd do something like your example, but I agree it's low risk and redundant to more rigorous mechanisms involving the type system.